### ISSUE TITLE
Pagination on page validate must redirect to right page not replace inside

### ISSUE THREAD

**Issue by DrDeath72** [yiisoft/yii2#11582]
wrong pages became copy of real last page

**Comment by samdark** [comment]
How to reproduce the issue?

**Comment by DrDeath72** [comment]
?page=99999999

**Comment by DrDeath72** [comment]
first version same behavior http://yiiframework.ru/news/page/100000

**Comment by samdark** [comment]
I see. So you want this to throw 404 instead, right?

**Comment by DrDeath72** [comment]
i think it must give you choice, 404, redirect or like now :) seo etc...

**Comment by tunecino** [comment]
isn't that a good thing ? I mean the last page is auto calculated by default while `yii\data\Pagination::getPageCount()` can be used if a different logic is required. same happens with REST: you get an `X-Pagination-Page-Count` value in headers or `_meta` to control it if needed. changing it now won't be a BC ?

**Comment by rob006** [comment]
> isn't that a good thing ?

This is definitely a bad thing if it is the default behavior. If page 123 returns data from page 99 it is simply a bug.

**Comment by githubjeka** [comment]
To fix this behavior set `\yii\data\Pagination::$validatePage` to `false`
By default it's `true`

>  * When this property is true, the value of [[page]] will always be between 0 and ([[pageCount]]-1).

**Comment by DrDeath72** [comment]
validatePage only link/alias wrong page to last

**Comment by githubjeka** [comment]
Yes it's by default and documented. To fix it you can set `validatePage` to `false`.

![image](https://cloud.githubusercontent.com/assets/874234/19142004/ca64773e-8ba2-11e6-951f-22d042997cb9.png)

**Comment by DrDeath72** [comment]
it's not equal 404 error, just empty data with wrong limit/offset

**Comment by githubjeka** [comment]
404 - No sure that it's correct when `$validatePage = false`. But when you set page set to a negative value or a word then the `page` param **always** return first page. This is bug sure.

**Comment by githubjeka** [comment]
And I don't understand responsibility of class Pagination. Is it should be to return NotFoundException eg...
Probably not. You should validate `page` param in controller and retrun NotFoundException youself.

**Comment by githubjeka** [comment]
@samdark Here it's necessary to clarify exactly what you need to improve.

**Comment by samdark** [comment]
Documentation at least.

**Comment by rob006** [comment]
> And I don't understand responsibility of class Pagination. Is it should be to return NotFoundException eg...
Probably - not. You should validate page param in controller and retrun NotFoundException youself.

IMO data widgets should do that. And this should be configurable.

And `\yii\data\Pagination::$validatePage` should be `true` by default.

**Comment by samdark** [comment]
I think we can remove the option in 2.1 since it doesn't make sense to have it false.

**Comment by samdark** [comment]
Not sure about changing default since it's a BC break.

**Comment by rob006** [comment]
IMO in 2.0.x data widgets should have an option to redirect or throw not found if there is no results on current page. This will be disabled by default and documented in 2.0.x, and enabled by default in 2.1.x.

**Comment by githubjeka** [comment]
You don't need to change anything right now.
`\yii\data\Pagination::$validatePage` is `true` by default.
All documented.

Issue about 404 and [bug](https://github.com/yiisoft/yii2/issues/11582#issuecomment-251875576)

**Comment by samdark** [comment]
OK so ideal solution is the one that gives you 404 if page isn't valid instead of falling back to 1st or last page.

**Comment by DrDeath72** [comment]
add options :D
-fallback
-redirect to right
-error404
-nothing

**Comment by rob006** [comment]
It should work in the same way as `UrlNormalizer::$action`, but without 301 redirection.
https://github.com/yiisoft/yii2/blob/master/framework/web/UrlNormalizer.php#L49-L69

**Comment by klimov-paul** [comment]
This is not a trivial thing to be resolved.

Redirect is not exactly a correct way of handling wrong page - it not better then current behavior as it indicates the requested page actually exists.
Redirect can not use 301 status as it will break the SEO as well: while at present state some 'blog' list has no page number 38, it may have it tomorrow, when new blog post will be added.

Showing 404 page is more logical - however such change breaks a BC. Also it is debatible: consider a filter form, which applies the search parameters to the `GET` - it also may produce several pages with different URLs but with the same content. Displaying first or last page on wrong page number is just the same issue. In general a canonical URL should be specified for any listing sensitive to the SEO.

Also, keep in mind that behavior of `yii\data\Sort` has same drawback - it also performs a fallback to defaults or ignore the sort in case sort specification is wrong (unexisting attribute, wrong format and so on).

Actually, page number as well as sorting is a part of user input. This can be seen pretty well in the scope of REST API, where these parameters are passed by API end user explicitly and not picked up from HTML links and buttons.
In order to deal with page number validation - its pick up should be placed inside the model - a search model, which creates a `DataProvider`. This model should perform a validation and populates itself with corresponding errors if any.
Remember, when user submits search form with incorrect inputs, like putting and words into a 'price' field - we show him an empty listing with search form, showing an error message. Wrong page number or sort specification can also be considered as a wrong input, which should be respond to with returning a proper error message to the user. For the REST API - this is actually an only option.

At the present state you can always validate page number yourself, if its validity is crucial for you. It is easy:

```php
if ($dataProvider->pagination->getPage == Yii::$app->request->getQueryParam('page')) {
   // valid
} else {
   // invalid
}
```

and deal with it as you pleased.
Or you can specify a canonical URL for your listing, excluding page, sort and filter parameters from it.

**Comment by klimov-paul** [comment]
I am afraid that putting a more automation into this matter will lead us to nowhere.
However, we may introduce an extra method `isValid()` to both `Pagination` and `Sort`. This method can be used to check, if pagination and sort parameters are valid or not, simplifying actual solution.

**Comment by rob006** [comment]
> Redirect is not exactly a correct way of handling wrong page - it not better then current behavior as it indicates the requested page actually exists.

Temporary redirection is much better than current behavior - it is compromise between UX and SEO. User still gets useful results, but you don't have duplicated content. 404 errors are worse - useless for end users and increasing 404 errors rate.

**Comment by rob006** [comment]
@terabytesoftw Are you sure you closed the right issue? Your comment seems to be completely unrelated to the problem described here.

**Comment by terabytesoftw** [comment]
> [@terabytesoftw](https://github.com/terabytesoftw) Are you sure you closed the right issue? Your comment seems to be completely unrelated to the problem described here.

Thanks, you're right, i posted the wrong issue.
